### ISSUE TITLE
iOS12 crashes in AMPDatabaseHelper

### ISSUE THREAD

**Issue by pmanuelli** [amplitude/Amplitude-iOS#171]
Hello everyone!

We're using Amplitude-iOS 4.3.1. and we're are having some crashes related to AMPDatabaseHelper in iOS 12 (I think it happen too in previous Amplitude-iOS version).

I can't reproduce any of this issues.
I leave attached the crashes reports.

Thanks in advance.
Pablo.

[inDatabaseWithStatement_issue.txt](https://github.com/amplitude/Amplitude-iOS/files/2390695/inDatabaseWithStatement_issue.txt)
[insertOrReplaceKeyValueToTable_issue.txt](https://github.com/amplitude/Amplitude-iOS/files/2390696/insertOrReplaceKeyValueToTable_issue.txt)
[addEventToTable_issue.txt](https://github.com/amplitude/Amplitude-iOS/files/2390697/addEventToTable_issue.txt)

**Comment by ghugues** [comment]
Hi,

We're experiencing the same issue. With version 4.3.1 on iOS 12 we're seeing a very important number of crashes in AMPDatabaseHelper relative to the number of users currently using iOS 12. We're expecting the total number of crashes to raise rapidly as users move to iOS 12.

We haven't reproduced this issue yet. I'm attaching two distinct crash reports from our crash reporting tool.
[30238166895338b2f7aeaa67eaf9528d.log](https://github.com/amplitude/Amplitude-iOS/files/2393791/30238166895338b2f7aeaa67eaf9528d.log)
[bc427aac364b0ae829147682473a3b98.log](https://github.com/amplitude/Amplitude-iOS/files/2393793/bc427aac364b0ae829147682473a3b98.log)

Thanks

**Comment by djih** [comment]
@pmanuelli @ghugues thanks for the report. I'm investigating the issue but not able to reproduce it myself. My initial hypothesis is that there are multiple threads trying to access the Sqlite database.

@ghugues in both of your crash reports, I am seeing at least 2 threads (one of which is Amplitude's SDK) trying to access Sqlite, and another maybe from your app or another library you are using.

@pmanuelli are you able to confirm if this is the case for you as well? Can you provide the full stack trace including other threads?

**Comment by ghugues** [comment]
@djih That was my initial hypothesis, but although it seems to happen a lot less often, I have examples of this crash without any other thread using Sqlite.
[ee05116bb5202034e2fb8502de1d62e2.log](https://github.com/amplitude/Amplitude-iOS/files/2393885/ee05116bb5202034e2fb8502de1d62e2.log)

**Comment by pmanuelli** [comment]
@djih Thanks for your answer. If I'm not wrong, the attachments that I provided earlier contains the full stack trace and includes other threads too.

Indeed, it seems that other threads are trying to access Sqlite too.

**Comment by djih** [comment]
@pmanuelli sorry you're right the full traces are there.

It's weird because our database helper hasn't changed much in the last 2-3 years, especially the logic around the queueing and enforcing single-threaded access to the database. I wonder if something changed in iOS 12 that affects and breaks this behavior.

If the issue is that multiple threads / libraries are trying to use Sqlite database, it's going to be extremely difficult to synchronize / serialize the access across all the threads. Within our own SDK we enforce it by queueing operations on a single thread, but we can't guard / coordinate with other libraries / modules trying to access Sqlite.

**Comment by ghugues** [comment]
It seems the example I gave you about a single thread using Sqlite might be a false positive. Although there is no `sqlite3_` function in the stack trace, one third party SDK is still in the process of doing Sqlite related work. So it is possible that the crash is still related to having multiple Sqlite databases opened at the same time.
Also, in the stacktraces @pmanuelli and I gave you, other third party SDKs include Firebase Analytics and Singular/Apsalar. This is going to be a problem because they are analytics SDKs, so they are very likely to be triggered simultaneously in a number of scenarios.

**Comment by ghugues** [comment]
Hi,
Any news on this issue ?

**Comment by pmanuelli** [comment]
Hi @ghugues.

We have not news about this issue. We've submitted a bug report to Apple about SQLite on iOS12, but have no answer so far.

As a workaround we have removed Singular from our app..

Maybe other workaround could be to put to delay between the initialization of the SDKs that access to SQLite and try to avoid the simultaneous access. Not a good solution, I know.

**Comment by ghugues** [comment]
@djih After further investigation, I've come to the same conclusion as @pmanuelli. The Apsalar might be the real culprit after all (the Apsalar SDK seem to be included in the Singular SDK).

I noticed that contrary to Firebase and Amplitude, the Apsalar SDK very rarely uses the database, so it was weird that it was always involved in crashes. It turns out that this SDK executes the following code during initialization:
```
if (sqlite3_threadsafe() >= 1) { // sqlite3_threadsafe() returns 2
    sqlite3_shutdown();
    sqlite3_config(SQLITE_CONFIG_SERIALIZED);
}
```
And the documentation says:
```
** The sqlite3_initialize() interface is threadsafe, but sqlite3_shutdown()
** is not.  The sqlite3_shutdown() interface must only be called from a
** single thread.  All open [database connections] must be closed and all
** other SQLite resources must be deallocated prior to invoking
** sqlite3_shutdown().
```
So if Amplitude is using SQLite at the same moment, it is probably what causes the crash. It's possible that previous versions of the library were more robust against such usage, but if this is indeed the source of these crashes, it's unlikely that Apple will fix anything as this is documented behavior.

However what makes this bug very likely to happen with the Amplitude SDK is that Amplitude uses `sqlite3_open` **a lot**. Actually, all database operations are wrapped by `AMPDatabaseHelper` between a `sqlite3_open` and a `sqlite3_close` call.
If you but a breakpoint on `sqlite3_open`, `sqlite3_open_v2`, `sqlite3_close` and `sqlite3_close_v2`, you will see that other frameworks that use SQLite (NSURLCache, CoreData, Google Analytics, Firebase, ...) don't wrap all their operations between an open and a close, they open the database once and never close it or only close it when they are definitely finished with it. So this is probably why this crash - although not caused by Amplitude - is much more likely to happen with Amplitude than any other SDK using SQLite.

I don't know much about SQLite, but how these frameworks do it suggests that it is a better practice to open the database once and keep it open. Maybe this is something that you could implement in your SDK ?

**Comment by ghugues** [comment]
@pmanuelli thanks. I should have mentioned that all of this is still guess work. I cannot remove the Apsalar SDK at the moment, so I con't confirm anything. Can you please report here if removing the SDK has solved the issue on your side ?

**Comment by pmanuelli** [comment]
@ghugues yes, in our case removing Singular SDK makes the crash disappear. But, of course, this is just a workaround.

**Comment by dimapv** [comment]
Hey guys, I'm from the mobile dev team at Singular - and we'll be releasing a fix shortly.

Thanks for the thorough investigation @pmanuelli & @ghugues, that code is indeed in our SDK and is our attempt at making sure that SQLite is running in thread-safe mode before using it.

We weren't aware of the unsafe nature of `sqlite3_shutdown` between threads and indeed up until now it didn't cause any trouble, but my guess is that apple changed some thread scheduling code in iOS12 and possibly caused this race-condition to bubble up.

We'll keep you posted with updates. Cheers :)

**Comment by rajusa420** [comment]
@ghugues - Thank you for posting the detailed analysis, we are hitting this crash as well but your analysis also made me realize that while setting the sqlite mode to SQLITE_CONFIG_MULTITHREAD was fine for our database access other SDKs in our app might depend on it being  SQLITE_CONFIG_SERIALIZED

**Comment by ghugues** [comment]
Hi,

If you are using SQLite in your app or in a third party SDK, I really recommend you read the official documentation about [Using SQLite In Multi-Threaded Applications](https://www.sqlite.org/threadsafe.html), as well as related ressources.

Let me make a recap of what I have read here and there since I started investigating this bug.

The SQLite library included in iOS is compiled with `SQLITE_THREADSAFE=2` (as reported by the `sqlite3_threadsafe()` function). I haven't tested in all versions of iOS and I encourage you to do so, but I'm pretty sure it will be the same. In this mode, by default, all you have to do to be safe is make sure that all calls using the same database connexion are serialized. The easiest way to do that is to use a GCD serial queue (like in the Amplitude SDK).
Note that if Apple ever compiles SQLite with `SQLITE_THREADSAFE=1` you will be in an even safer mode so you have nothing to worry about, and it's very unlikely that they will ever compile it with `SQLITE_THREADSAFE=0`, that would make no sense on a multitasking OS like iOS.

If for some reason you need the serialized mode (for example, because you have legacy code that accesses the same database connexion from multiple threads without serialization), then you have two options :
 - Use `sqlite3_open_v2` with the `SQLITE_OPEN_FULLMUTEX` option. This is, in my opinion, the preferred option.
 - Call `sqlite3_config(SQLITE_CONFIG_SERIALIZED)` when the app starts. Because this call is not thread safe, it should be done as early as possible and must be done before calling `UIApplicationMain`. If you're a third party SDK developer, you can't do that yourself, you need to ask the developer to do it themselves when they integrate your SDK.

Never, under any circumstance, use `sqlite3_config` or `sqlite3_shutdown` from a third party SDK, you have no way of making this work safely. In the best case scenario you will create a latent bug like in the Apsalar SDK (I would actually argue that this is the worst case scenario).

**Comment by pmanuelli** [comment]
@dimapv thank you! I will try this new version and let you know if the issue has gone.

**Comment by numist** [comment]
In general, using `sqlite3_config` from library code is not recommended as it affects all other code in the process. Its use should be reserved for the top layer, where no one else can be blamed (or hurt).

Many global options have [connection-local analogues](https://www.sqlite.org/c3ref/c_dbconfig_enable_fkey.html) set via [`sqlite3_db_config`](https://www.sqlite.org/c3ref/db_config.html).
